### PR TITLE
Expose configuration properties to configure the  WebsocketServerSpec to use

### PR DESCRIPTION
**Pull request by polarbear567**
For issue [29550](https://github.com/spring-projects/spring-boot/issues/29550).
Preliminary submission.

### CODE REVIEW

**Comment by philwebb** [comment]
Thanks for the pull-request @polarbear567. When raising #29550 I was originally thinking we'd offer a way to do this programmatically. The properties are an interesting alternative that didn't occur to me and I'm now not sure which is best. I'm going to flag this for team discussion to see what other think.

**Comment by wilkinsona** [comment]
I quite like the property-based approach. They do feel quite low-level but not so low that I think they ought to only be available programatically.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketProperties.java [hunk 1]
```diff
@@ -121,6 +123,74 @@ public void setSsl(Ssl ssl) {
...
+		}
+
+		public void setSpec(Spec spec) {
+			this.spec = spec;
+		}
+
+		public static class Spec {
+
+			/**
+			 * Sub-protocol to use in websocket handshake signature. Null by default.
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketProperties.java:137]: There's no need to write "by default" in the description as we harvest defaults and process them separately.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketProperties.java [hunk 2]
```diff
@@ -121,6 +123,74 @@ public void setSsl(Ssl ssl) {
...
+			 */
+			private String protocols;
+
+			/**
+			 * Specifies a custom maximum allowable frame payload length. 65536 by
+			 * default.
+			 */
+			private int maxFramePayloadLength;
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketProperties.java:145]: This field should be initialized  with `65536` then. Currently the advertised default is `0`.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketProperties.java [hunk 3]
```diff
@@ -121,6 +123,74 @@ public void setSsl(Ssl ssl) {
...
+
+			/**
+			 * Flag whether to proxy websocket ping frames or respond to them. False by
+			 * default.
+			 */
+			private boolean handlePing;
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketProperties.java:151]: The second sentence should be removed as discussed above.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketProperties.java [hunk 4]
```diff
@@ -121,6 +123,74 @@ public void setSsl(Ssl ssl) {
...
+
+			/**
+			 * Flag whether the websocket compression extension is enabled if the client
+			 * request presents websocket extensions headers. By default compression is
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketProperties.java:155]: Ditto.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketWebSocketNettyRouteProvider.java [hunk 5]
```diff
@@ -42,19 +45,35 @@
...
-	RSocketWebSocketNettyRouteProvider(String mappingPath, SocketAcceptor socketAcceptor,
+	private final String protocols;
+
+	private final int maxFramePayloadLength;
+
+	private final boolean handlePing;
+
+	private final boolean compress;
+
+	RSocketWebSocketNettyRouteProvider(String mappingPath, Spec spec, SocketAcceptor socketAcceptor,
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketWebSocketNettyRouteProvider.java:56]: There's no need to assign the fieds here. keeping the `Spec` object should be enough.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketWebSocketNettyRouteProvider.java [hunk 6]
```diff
@@ -42,19 +45,35 @@
...
 			Stream<RSocketServerCustomizer> customizers) {
 		this.mappingPath = mappingPath;
 		this.socketAcceptor = socketAcceptor;
 		this.customizers = customizers.collect(Collectors.toList());
+		this.protocols = spec.getProtocols();
+		this.maxFramePayloadLength = (spec.getMaxFramePayloadLength() <= 0) ? 65536 : spec.getMaxFramePayloadLength();
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/rsocket/RSocketWebSocketNettyRouteProvider.java:62]: This handling of the default here doesn't feel right to me.

**Review by snicoll — CHANGES_REQUESTED** [review]
Thanks for the PR @polarbear567. We need tests that exercise those properties and validate they are mapped properly please. Would you have time to look at that?

**Comment by polarbear567** [comment]
Sure, I'll take a look at it when I'm free. I'll modify PR according to comments. I don't have any good ideas for testing. Do you have any suggestions? @snicoll

**Comment by snicoll** [comment]
> I don't have any good ideas for testing. Do you have any suggestions?

Yeah good point, I didn't look far enough. The mapping is hidden behind the predicate so it's hard to test this way. We can look ourselves as part of merging this, thanks for the update!

**Pull request closed without merging**